### PR TITLE
Air alarm & central atmos computer & vents & scrubbers unatomic featurecreep bug-a-palooza

### LINKED ISSUE #15023: The Xenobiology Atmospheric Computer on Box Station becomes a Central Atmospheric Computer with this one, simple trick.
The Xenobiology Atmospheric Computer becomes a Central Atmospheric Computer if you simply screwdriver the screen out and screwdriver it back in.  This is capable of controlling all air alarms on the station.

### LINKED ISSUE #17707: Central Atmospherics Computer Doesn't Function Correctly
#### Description of issue
The Central Atmospherics computer does not respond when clicking on individual air alarms, when changing Z levels, or when showing a detailed list. 

#### Difference between expected and actual behavior
When clicking on an individual air alarm, it brings it up, and allows the modification of values in the alarm. 

When selecting a new Z level, a new Z level should load with its map. 

When clicking on show detail list, a detail list should appear. 

#### Steps to reproduce 
Click on the CAC
Click on an air alarm
Wait 10 minutes
Nothing happens

Click on the CAC
Click on change Z level
Select a Z level
Wait 10 minutes
Nothing Happens

Click on the CAC 
Click on Show Detailed List
Wait 10 minutes
Nothing Happens. 

#### Length of time in which bug has been known to occur
Since https://github.com/vgstation-coders/vgstation13/issues/16654 probably. 
ahhhhhh

#### Server revision
Current

#### Issue bingo

Please check whatever applies. More checkboxes checked increase your chances of not being yelled at by every contributor.

- [x] Issue could be reproduced at least once
- [x] Issue could be reproduced by different players
- [x] Issue could be reproduced in multiple rounds
- [x] Issue happened in a recent (less than 7 days ago) round
- [ ] [Couldn't find an existing issue about this](https://github.com/d3athrow/vgstation13/issues)

### LINKED ISSUE #21424: Feature request: [Export air alarms settings to room] button
#### Description of issue

That or an actual notepad import/export feature like what the AACs have.
Being able to set one air alarm and then press the button to set the other 70 air alarms in the same hallway that for some reason don't copy the settings would be a godsend for both rogue and non-rogue atmos memery.

[featurerequest]

### PR DESCRIPTION
**Pull request by brndd**
[bugfix] [gameplay] [tweak] [qol]

What started out as a venture to add a sort of a batch editing functionality to the Central Atmospherics Computer turned into a bugfixing and general improvement rampage on everything related to air alarms.

Misc bugfixes transcribed from my notes (it's possible I missed some because there were lots):

- Most settings (everything I could think of) on air alarms now propagate to other air alarms in the area
- There's a manual override button on air alarms to force the firelocks in the area open. This lasts until the override is canceled and has no special indication that it's been done, so there may be some improvements left to be made to this.
- Air alarm thresholds have a reset button.
- Air alarm UI is now a little wider so it has no horizontal scrollbar.
- Unlocked air alarms aren't bound by the thermostat cap (but their heating and cooling capabilities are still limited between -40°C and 90°C).
- Related to above, thermostats no longer cap out at -40°C and 90°C, they just won't cool/heat any lower/higher than those values. This stops air alarms from trying to heat up server rooms (and miserably failing to do so).
- Thermostat cap is now inclusive and not exclusive. The heating/cooling stops within 0.5 degrees of the cap anyway so it won't trigger the alert by heating the room too much.
- Firelocks that are manually opened just before an alarm clears no longer automatically close back down after the alarm has cleared.
- Fixed the vent pressure check buttons in the air alarm GUI
- Fixed the vent blowing/sucking (or siphoning/releasing or purging/stabilizing depending on which of the SIX DIFFERENT TERMS we had for this you preferred) buttons on air alarms.
- Vents now consistently use the words "siphon" and "blow" for sucking and blowing respectively. "Blow" is up to debate, it used to be called "release" (and in some other areas these were purge/stabilize).
- Air alarm UI now updates immediately when locked/unlocked.
- Air alarm UI remote control now works (it used to do nothing). "Auto" mode has been removed, it's now a Yes/No thing. Remote control can only be toggled locally, not from the CAC, and requires unlocking the AA to toggle.
- Air alarm UI now has an internal pressure checks setting for vents.
- Fixed a bug that stopped the CAC tracker map from working when an air alarm was selected in the other view.
- Added a button that lets you choose whether a room cycles or not after changing the mode at an air alarm.
- Added a special air alarm for the execution chamber in the brig. Didn't map it in to avoid mapping gas. Yes, this means that the CAC in the brig execution chamber is broken for the time being (you can't use it to control the air alarm as sec because you have no access). This is probably no biggie because nobody fucking uses the gas chamber.

**Central Atmospherics Computer:**

- Air alarm presets system overhauled:
  - There are now 4 default presets: Human, Vox, Coldroom and Plasmaman (which exists totally for making habitable areas for plasmamen and not as a base for plasmafloods)
  - Custom presets can be added and removed at the CAC with CE access/emag. Default presets can't be removed, added or renamed, but can be modified.
  - Presets can control the various thresholds, target temperature, target pressure (this is actually the average of the middle two pressure thresholds), and which gases are scrubbed by scrubbers.
  - CE access/emag can apply presets to all air alarms on the station as a batch, to eg. switch all Human presets over to Vox.
  - Same access can also edit presets to eg. switch all Human presets to scrub for N2O.
- CAC access system overhauled. You now log into it similarly to the Communications Console by holding a card and clicking the Log In button. You'll stay logged in until you log out again, so that you can forget to log out your full access and some assistant fucks with the terminal. Emag bypasses the log in.
- CAC now only lets you control air alarms that you have the access to control. For instance logging in with xenobiology access lets you control xenobio air alarms and nothing else, regardless of which CAC you're using.
- CAC now respects the remote control setting of air alarms. Note that this is not the same as AI control.

The intention of this preset system is to make AAC script fuckery less necessary, and easier to deal with without having pre-prepared scripts of your own.

A quick webum showing the preset switch in action: https://streamable.com/542r2

The code here isn't pretty and I'm not proud of much it, but what I built upon wasn't pretty either so I kind of had to make do with what I had. Didn't want to refactor every single thing.

Closes #21424, closes #17707 (can not reproduce), closes #15023. Doesn't cl\*se #7616 because I couldn't figure out a way to do it: I suspect radio signals are async or something.

:cl:
 * rscadd: There's a new presets system in the Central Atmospherics Computer that lets you, among other things, set all air alarms on station to Vox or Plasmaman mode.
 * bugfix: Fixed a metric fuckton of bugs related to air alarms, the Central Atmospherics Computer, etc.
 * tweak: Air alarms now have a button that forces all the firelocks in the area open. This button is only accessible when the air alarm is unlocked.

### CODE REVIEW

**Comment by brndd** [comment]
Oh yeah and I moved the file out of the WorkInProgress directory because it's been merged into the game since 2012 that is seven fucking years.

**Comment by SonixApache** [comment]
- Most settings (everything I could think of) on air alarms now propagate to other air alarms in the area

Please make this a button. I don't have a major preference for a "do this" or "don't do this" button, but it should be a button regardless.
- There's a manual override button on air alarms to force the firelocks in the area open. This lasts until the override is canceled and has no special indication that it's been done, so there may be some improvements left to be made to this.

Definitely add some sort of indicator to this. Maybe toss it in the yellow [atmos alarm in this area]  bar?

**Comment by brndd** [comment]
> Please make this a button. I don't have a major preference for a "do this" or "don't do this" button, but it should be a button regardless.

I don't really see the point though. Why would you ever *not* want other air alarms that are in control of the same area to share the same safety threshold or thermostat settings, for instance?

> Definitely add some sort of indicator to this. Maybe toss it in the yellow [atmos alarm in this area] bar?

I mean, there's an indicator in the UI (the button turns green to indicate it's enabled and says "Disable manual override"), and the alarm sprites will keep blinking red (or yellow). It's just not blatantly obvious to passers-by that the fire doors have been forced open and that the area could potentially be a vacuum. I thought about adding the blinking red alarm overlay (that happens with eg. shuttle calls and fire alarms too) to it, but that's already used for other things so it might not help.

**Comment by Hoshi-chan** [comment]
Just remove firelocks, tbh.

**Comment by ghost** [comment]
I'm not sure you realize how lethal it is to set all of stations scrubbers to scrub O2 at the same time. The new plasmaman preset will make it trivial to scrub all air (nitrogen included) from entire station with one button and no preparation needed.

I'm not opposed to any of this however, since I can't wait to abuse it myself.

**Comment by FudgePucker** [comment]
Random idea I had was to make it so that emag'd air alarms dont alert, or even check the pressure or tempature. In fact it should be obvious if you look at the air alarm GUI, but at a glance it looks normal.

Pressure: cozy
Tempature: fine
Happiness: 20%
Sunshine: 80%
CO2: 0%
N2O: 0%
Toxins: 0%

**Comment by brndd** [comment]
> I'm not sure you realize how lethal it is to set all of stations scrubbers to scrub O2 at the same time. The new plasmaman preset will make it trivial to scrub all air (nitrogen included) from entire station with one button and no preparation needed.

Oh hoh there's something even more lethal in the pipeline that I'm hopefully fixing soon (thankfully not related to this PR and not even possible through my interface). It takes several minutes for scrubbers to actually get all the air in a room out. Either way you can already do this with an AAC script.

**Comment by brndd** [comment]
> 
> 
> Random idea I had was to make it so that emag'd air alarms dont alert, or even check the pressure or tempature. In fact it should be obvious if you look at the air alarm GUI, but at a glance it looks normal.
> 
> Pressure: cozy
> Tempature: fine
> Happiness: 20%
> Sunshine: 80%
> CO2: 0%
> N2O: 0%
> Toxins: 0%

That's a pretty good idea, but I'll probably stick it in another PR because this is already bloated enough.

**Comment by ghost** [comment]
Scripts can't alter air alarm tolerances.

**Comment by brndd** [comment]
> Scripts can't alter air alarm tolerances.

True, they can't, but scripts can make all the scrubbers suck the air out and firedoors falling down isn't going to stop that from happening

**Comment by SonixApache** [comment]
>Why would you ever not want other air alarms that are in control of the same area to share the same safety threshold or thermostat settings, for instance?

When there's a breach, and you want to turn off the atmos shit while you fix it. And on the off chance that you make a small different-atmos area without having access to area defining tools.

**Comment by brndd** [comment]
>When there's a breach, and you want to turn off the atmos shit while you fix it.

I don't see how *not* mirroring configuration to all the air alarms in the area would be beneficial here. They all control the same vents and scrubbers, and more than one air alarm in an area effectively just provides two things: extra gas sensors and an extra terminal for human access. Pick any one of the 6 air alarms in the central hallway on box and you can control every single vent and scrubber in that hallway from it, and all the Mode settings will apply to every vent and scrubber.

There's one situation where not mirroring the settings has an effect and that's the thermostat. Currently if you set eg. the top air alarm in this picture to 15°C and the bottom one to 25°C, the top one will try to cool the hallway while the bottom one will try to heat it up. I fail to see any practical application for this, because 1) the way our atmos works means this won't create localized pockets of hot/cool air in the hallway, and 2) nobody uses the thermostat anyway ever.




> And on the off chance that you make a small different-atmos area without having access to area defining tools.

This won't help. Air alarms affect every vent and scrubber in the area (as defined by eg. station blueprints) they're in. You can't make a "small different-atmos area" like that, and this will in fact just confuse the air alarm if the different-atmos area happens to have vastly different atmospheric conditions.

Concrete example: the execution chamber. There are two separate ZAS areas here, the gas chamber itself and the rest of the room. The air alarm is in the rest-of-the-room part, and has zero idea what's happening in the gas chamber. N2O in the gas chamber won't trigger the alarm, low pressure in the gas chamber won't trigger it, etc. Now if you were to put the air alarm *in* the gas chamber part, it would think the entire room is the same as the gas chamber. There's N2O in the gas chamber but not the rest of the room? The entire room is N2O, shut the firelocks.

If you were to have two air alarms here, the alert system will always pick the one with the highest alarm level (in this case usually the one inside of the gas chamber) and transfer its alarm to the whole room. This is how it already works, it's not something I added (and it's really the only way it *could* work.)

EDIT: After rereading this thread I think I realised what you meant: having an air alarm in the gas chamber and setting its thresholds to not trigger on N2O. This is something that's currently possible because settings don't propagate, and this PR would remove that possibility. I'm honestly not sure if this is ever done (it certainly isn't this way on any map at roundstart), and because of that I don't think such a marginal use-case is worth the extra UI clutter of adding in yet another button. But I can put in a button if you really, *really* think it's worthwhile.

File: code/game/machinery/computer/atmos_control.dm [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+
+/obj/machinery/computer/atmoscontrol/attackby(var/obj/item/I as obj, var/mob/user as mob)
+	if(istype(I, /obj/item/weapon/card/emag))
+		emag(user)
+	return ..()
+
+/obj/machinery/computer/atmoscontrol/emag(var/obj/item/emag as obj, var/mob/user as mob)
+	if(!emagged)
+		emagged = 1
+		new/obj/effect/effect/sparks(get_turf(src))
```
Comment (Kurfursten) [inline code/game/machinery/computer/atmos_control.dm]: use a spark system, if my memory serves the sparks will never get deleted if you do it this way

File: code/game/machinery/computer/atmos_control.dm [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+		return 1
+
+	if(href_list["close"])
+		if(usr.machine == src)
+			usr.unset_machine()
+		return 1
+
+	if(href_list["reset"])
+		current = null
+		return TRUE
```
Comment (Kurfursten) [inline code/game/machinery/computer/atmos_control.dm]: weird inconsistency that you swap between 1 and TRUE

Comment (brndd) [reply code/game/machinery/computer/atmos_control.dm]: The ones that use TRUE are legacy code while the ones that use 1 were written by me.

File: code/game/machinery/computer/atmos_control.dm (hunk 1 above)
Comment (SonixApache) [reply code/game/machinery/computer/atmos_control.dm]: Funny enough, I was about to report this today. Someone should really fix that.

File: code/game/machinery/computer/atmos_control.dm (hunk 2 above)
Comment (Kurfursten) [reply code/game/machinery/computer/atmos_control.dm]: that's fine just pick one direction and consistify them

File: code/game/machinery/computer/atmos_control.dm (hunk 1 above)
Comment (brndd) [reply code/game/machinery/computer/atmos_control.dm]: `spark(src, 1, FALSE)` gonna work?

File: code/game/machinery/computer/atmos_control.dm (hunk 2 above)
Comment (brndd) [reply code/game/machinery/computer/atmos_control.dm]: I went with 1 because that's what nanouis in general use.

**Comment by brndd** [comment]
Air alarms now apply their preset mode when they're first created. This won't have much of an effect in-game (or shouldn't), but will help when mapping because now you don't need to use special scrubbers to eg. create vox areas, just a special air alarm.

**Pull request merged**